File: py/513.找树左下角的值.py
#
# @lc app=leetcode.cn id=513 lang=python3
#
# [513] 找树左下角的值
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def findBottomLeftValue(self, root: TreeNode) -> int:
        trace = [root]
        cur_trace = []
        re = root.val
        while trace:
            for node in trace:
                if node.left or node.right:
                    if node.left:
                        cur_trace.append(node.left)
                    if node.right:
                        cur_trace.append(node.right)
            
            if cur_trace == []:
                trace = []
            if cur_trace:
                re = cur_trace[0].val
                trace = cur_trace
                cur_trace = []

        return re
# @lc code=end
root = TreeNode(2)
n1 = TreeNode(1)
n2 = TreeNode(3)
n3 = TreeNode(4)
n4 = TreeNode(5)
n5 = TreeNode(6)
root.left = n1
root.right = n2
n2.left = n3
n2.right = n4
print(Solution().findBottomLeftValue(root))

File: golang/513.找树左下角的值.go
// package main

/*
 * @lc app=leetcode.cn id=513 lang=golang
 *
 * [513] 找树左下角的值
 */

// @lc code=start

// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func findBottomLeftValue(root *TreeNode) (re int) {
	re = root.Val
	trace := []*TreeNode{}
	cur_trace := []*TreeNode{}
	trace = append(trace, root)
	for len(trace) > 0 {
		for i := 0; i < len(trace); i++ {
			if trace[i].Left != nil {
				cur_trace = append(cur_trace, trace[i].Left)
			}
			if trace[i].Right != nil {
				cur_trace = append(cur_trace, trace[i].Right)
			}
		}
		if len(cur_trace) != 0 {
			re = cur_trace[0].Val
		}
		trace = []*TreeNode{}
		for i := 0; i < len(cur_trace); i++ {
			trace = append(trace, cur_trace[i])
		}
		cur_trace = []*TreeNode{}
	}
	return
}

// @lc code=end
// func main() {
// 	node1 := TreeNode{
// 		Val: 1,
// 	}
// 	node2 := TreeNode{
// 		Val: 3,
// 	}
// 	node3 := TreeNode{
// 		Val: 3,
// 	}
// 	node4 := TreeNode{
// 		Val: 4,
// 	}
// 	node5 := TreeNode{
// 		Val: 5,
// 	}
// 	node6 := TreeNode{
// 		Val: 6,
// 	}
// 	node1.Left = &node2
// 	node1.Right = &node3
// 	node2.Left = &node4
// 	node3.Right = &node5
// 	node5.Left = &node6
// 	fmt.Println(findBottomLeftValue(&node1))
// }
